presto-orc/src/test/java/com/facebook/presto/orc/OrcTester.java

```diff
@@ -478,28 +485,75 @@ private void testRoundTripType(Type type, List<?> readValues)
+            throws Exception
     {
         assertEquals(types.size(), readValues.size());
+        Optional<OrderChecker> optionalOrderChecker = Optional.empty();
+
+        if (order.isPresent()) {
+            OrderChecker orderChecker = new OrderChecker();
+            orderChecker.setDesiredOrder(order.get());
+            // Turn normal TupleDomainFilters into their corresponding OrderTrackingOrcFilters.
+            // Keeping types the same so that we can
+            // 1. reuse the "filters" variable;
+            List<Map<Integer, Map<Subfield, TupleDomainFilter>>> newFilters = filters.stream()
+                    .map(filter -> Maps.transformEntries(filter,
+                            (column, value) -> Maps.transformValues(value,
+                                    tupleDomainFilter -> addOrderTracking(tupleDomainFilter, orderChecker, column))))
+                    .collect(toImmutableList());
+            optionalOrderChecker = Optional.of(orderChecker);
+            filters = newFilters;
+        }
 
         // forward order
-        assertRoundTrip(types, readValues, filters);
+        assertRoundTrip(types, readValues, filters, optionalOrderChecker);
 
         // reverse order
         if (reverseTestsEnabled) {
-            assertRoundTrip(types, Lists.transform(readValues, OrcTester::reverse), filters);
+            assertRoundTrip(types, Lists.transform(readValues, OrcTester::reverse), filters, Optional.empty());
         }
 
         if (nullTestsEnabled) {
             // forward order with nulls
-            assertRoundTrip(types, insertNulls(types, readValues), filters);
+            assertRoundTrip(types, insertNulls(types, readValues), filters, optionalOrderChecker);
 
             // reverse order with nulls
             if (reverseTestsEnabled) {
-                assertRoundTrip(types, insertNulls(types, Lists.transform(readValues, OrcTester::reverse)), filters);
+                assertRoundTrip(types, insertNulls(types, Lists.transform(readValues, OrcTester::reverse)), filters, optionalOrderChecker);
             }
         }
     }
 
+    /*
+        This is a syntactic sugar for how we extended testRoundTripTypes. We added this to make the lives of the test-writer easier. (By forcing orders not to be null).
+     */
+    public void testRoundTripTypesWithOrder(List<Type> types, List<List<?>> readValues, List<Map<Integer, Map<Subfield, TupleDomainFilter>>> filters, List<Integer> order)
+            throws Exception
+    {
+        assertNotNull(order);
+        testRoundTripTypes(types, readValues, filters, Optional.of(order));
+    }
+
+    private static TupleDomainFilter addOrderTracking(TupleDomainFilter filter, OrderChecker orderChecker, int column)
+    {
+        if (filter instanceof BigintRange) {
+            OrderTrackingOrcFilter newFilter = TestBigintRange.of((BigintRange) filter, orderChecker);
+            newFilter.register(column);
```

Comment:
- remove `register` method and pass `column` to the constructor
- also, a better way could be to pass a callback that should be invoked on every `testXxx` method; this way we can have `TrackingTupleDomainFilter` decoupled from `OrderChecker`, e.g. it won't have to know about what's being tracked.

```
        if (filter instanceof BigintRange) {
            return TestBigintRange.of((BigintRange) filter, orderChecker, column);
        }
        if (filter instanceof DoubleRange) {
            return TestDoubleRange.of((DoubleRange) filter, orderChecker, column);
        }
```

or

```
public interface TrackingTupleDomainFilter
        extends TupleDomainFilter
{
    class TupleDomainFilterOrderChecker
    {
        private final List<Integer> expectedOrder;
        private final List<Integer> columnOrdering = new ArrayList<>();

        TupleDomainFilterOrderChecker(List<Integer> expectedOrder)
        {
            this.expectedOrder = requireNonNull(expectedOrder, "expectedOrder is null");
        }

        public void call(int column)
        {
            if (!columnOrdering.contains(column)) {
                columnOrdering.add(column);
            }
        }

        public void assertOrder()
        {
            assertEquals(columnOrdering, expectedOrder);
        }
    }

    class TestBigintRange
            extends BigintRange
            implements TrackingTupleDomainFilter
    {
        private final Consumer<TupleDomainFilter> onTestCallback;

        public static TestBigintRange of(BigintRange inRange, Consumer<TupleDomainFilter> onTestCallback)
        {
            return new TestBigintRange(inRange.getLower(), inRange.getUpper(), inRange.nullAllowed, onTestCallback);
        }

        private TestBigintRange(long lower, long upper, boolean nullAllowed, Consumer<TupleDomainFilter> onTestCallback)
        {
            super(lower, upper, nullAllowed);
            this.onTestCallback = requireNonNull(onTestCallback, "onTestCallback is null");
        }

        @Override
        public boolean testLong(long value)
        {
            onTestCallback.accept(this);
            return super.testLong(value);
        }
    }

    class TestDoubleRange
            extends DoubleRange
            implements TrackingTupleDomainFilter
    {
        private final Consumer<TupleDomainFilter> onTestCallback;

        public static TestDoubleRange of(DoubleRange inRange, Consumer<TupleDomainFilter> onTestCallback)
        {
            return new TestDoubleRange(inRange.lower, inRange.lowerUnbounded, inRange.lowerExclusive, inRange.upper, inRange.upperUnbounded, inRange.upperExclusive, inRange.nullAllowed, onTestCallback);
        }

        private TestDoubleRange(double lower, boolean lowerUnbounded, boolean lowerExclusive, double upper, boolean upperUnbounded, boolean upperExclusive, boolean nullAllowed, Consumer<TupleDomainFilter> onTestCallback)
        {
            super(lower, lowerUnbounded, lowerExclusive, upper, upperUnbounded, upperExclusive, nullAllowed);
            this.onTestCallback = requireNonNull(onTestCallback, "onTestCallback is null");
        }

        @Override
        public boolean testDouble(double value)
        {
            onTestCallback.accept(this);
            return super.testDouble(value);
        }
    }
}


    private static TupleDomainFilter addOrderTracking(TupleDomainFilter filter, TupleDomainFilterOrderChecker orderChecker, int column)
    {
        if (filter instanceof BigintRange) {
            return TestBigintRange.of((BigintRange) filter, unused -> orderChecker.call(column));
        }
        if (filter instanceof DoubleRange) {
            return TestDoubleRange.of((DoubleRange) filter, unused -> orderChecker.call(column));
        }
        throw new UnsupportedOperationException("Unsupported filter type: " + filter.getClass().getSimpleName());
    }


        Optional<Map<Integer, Map<Subfield, TupleDomainFilter>>> transformedFilters = filters;
        if (orderChecker.isPresent()) {
            transformedFilters = filters.map(
                    filter -> Maps.transformEntries(filter,
                            (column, value) -> Maps.transformValues(value,
                                    tupleDomainFilter -> addOrderTracking(tupleDomainFilter, orderChecker.get(), column))));
        }
        try (OrcSelectiveRecordReader recordReader = createCustomOrcSelectiveRecordReader(file, orcEncoding, orcPredicate, types, MAX_BATCH_SIZE, transformedFilters.orElse(ImmutableMap.of()), filterFunctions, filterFunctionInputMapping, requiredSubfields)) {
            assertEquals(recordReader.getReaderPosition(), 0);
            ...
        }
        orderChecker.ifPresent(TupleDomainFilterOrderChecker::assertOrder);


```

Reply:
I get this. This is cool! Fun how a lambda can turn into a Consumer. Is that a Java 8 syntactic sugar thing, or am I misunderstanding something?

Reply:
I get this. This is cool! Fun how a lambda can turn into a Consumer. Is that a Java 8 syntactic sugar thing, or am I misunderstanding something?